File: src/util/project.ts
import type { Config } from "../config.js";
import type { FieldValue, ProjectField, ProjectGraph, ProjectItem } from "../types.js";

/** Look up a project field by (case-insensitive) name. */
export function findField(graph: ProjectGraph, name: string | undefined): ProjectField | undefined {
  if (!name) return undefined;
  const lower = name.toLowerCase();
  return graph.fields.find((f) => f.name.toLowerCase() === lower);
}

/** Require a field to exist, throwing a helpful error if the config points at a missing field. */
export function requireField(graph: ProjectGraph, name: string, purpose: string): ProjectField {
  const field = findField(graph, name);
  if (!field) {
    const available = graph.fields.map((f) => f.name).join(", ");
    throw new Error(`Field "${name}" (needed for ${purpose}) not found on project. Available fields: ${available}`);
  }
  return field;
}

function valueFor(item: ProjectItem, fieldName: string): FieldValue | undefined {
  const lower = fieldName.toLowerCase();
  return item.fieldValues.find((v) => v.fieldName.toLowerCase() === lower);
}

export function statusOf(item: ProjectItem, cfg: Config): string | undefined {
  return valueFor(item, cfg.fields.status)?.singleSelect?.name;
}

/** When the Status field value was last changed — our proxy for "time in column". */
export function statusUpdatedAt(item: ProjectItem, cfg: Config): string | undefined {
  return valueFor(item, cfg.fields.status)?.updatedAt;
}

/** The item's Priority option name, or undefined if unset. */
export function priorityOf(item: ProjectItem, cfg: Config): string | undefined {
  return valueFor(item, cfg.fields.priority)?.singleSelect?.name;
}

/** The iteration the item is assigned to (title + id), or undefined if unset. */
export function iterationOf(item: ProjectItem, cfg: Config): { title: string; iterationId: string } | undefined {
  return valueFor(item, cfg.fields.iteration)?.iteration;
}

/** The item's Estimate value, or undefined if no estimate field is configured or set. */
export function estimateOf(item: ProjectItem, cfg: Config): number | undefined {
  if (!cfg.fields.estimate) return undefined;
  return valueFor(item, cfg.fields.estimate)?.number;
}

/** True when the item's status is one of `doneStatuses` (case-insensitive). */
export function isDone(item: ProjectItem, cfg: Config): boolean {
  const status = statusOf(item, cfg);
  if (!status) return false;
  return cfg.doneStatuses.some((d) => d.toLowerCase() === status.toLowerCase());
}

/** Resolve a single-select option id by option name (case-insensitive). */
export function optionId(field: ProjectField, name: string): string | undefined {
  const lower = name.toLowerCase();
  return field.options?.find((o) => o.name.toLowerCase() === lower)?.id;
}

File: ruby/lib/boardly/util/project.rb
# frozen_string_literal: true

module Boardly
  module Util
    # Accessors that map configured field NAMES to values on an item, plus small
    # field lookups. Mirrors the TS `util/project.ts`.
    module Project
      module_function

      def find_field(graph, name)
        return nil if name.nil?

        lower = name.to_s.downcase
        graph.fields.find { |f| f.name.to_s.downcase == lower }
      end

      def require_field(graph, name, purpose)
        field = find_field(graph, name)
        return field if field

        available = graph.fields.map(&:name).join(", ")
        raise Boardly::ConfigError, %(Field "#{name}" (needed for #{purpose}) not found on project. Available fields: #{available})
      end

      def value_for(item, field_name)
        lower = field_name.to_s.downcase
        item.field_values.find { |v| v.field_name.to_s.downcase == lower }
      end

      def status_of(item, cfg) = value_for(item, cfg.fields[:status])&.single_select&.name

      def status_updated_at(item, cfg) = value_for(item, cfg.fields[:status])&.updated_at

      def priority_of(item, cfg) = value_for(item, cfg.fields[:priority])&.single_select&.name

      def iteration_of(item, cfg) = value_for(item, cfg.fields[:iteration])&.iteration

      def estimate_of(item, cfg)
        return nil unless cfg.fields[:estimate]

        value_for(item, cfg.fields[:estimate])&.number
      end

      def done?(item, cfg)
        status = status_of(item, cfg)
        return false unless status

        cfg.done_statuses.any? { |d| d.downcase == status.downcase }
      end

      def option_id(field, name)
        lower = name.to_s.downcase
        opt = (field.options || []).find { |o| o[:name].to_s.downcase == lower }
        opt && opt[:id]
      end
    end
  end
end
